### PR TITLE
feat: make preview input loading lightweight

### LINKED ISSUE #28: feat: make preview input loading lightweight
### Priority Level

Medium

### Task Summary

Make data loading in `preview()` lightweight. 

Right now `preview()` appears to load the full input dataset up front and only limits later stages via `preview_num_records`. For larger inputs, that makes preview heavier than expected for a fast inspection path.

### Technical Details & Implementation Plan

We want `preview()` to behave like a genuinely lightweight path end to end, not just in downstream pipeline execution.

Today, the preview flow appears to materialize the full input dataset before `preview_num_records` takes effect in later stages. That means preview can still incur the full cost of reading large inputs even when only a small subset of rows is actually processed.

The goal of this task is to change that contract so preview truncates earlier during input loading.

Proposed work:
- Update the preview path so row limiting happens during input loading rather than after the full dataset has already been materialized
- Thread the preview row limit into the input reader layer
- Keep the `run()` path unchanged
- Update preview logging appropriately
- Add tests covering inputs larger than the requested preview size

Likely affected areas:
- `src/anonymizer/interface/anonymizer.py`
- `src/anonymizer/engine/io/reader.py`
- any format-specific reader helpers used by `AnonymizerInput`
- preview/logging tests

### Dependencies

_No response_

### PR DESCRIPTION
**Pull request by memadi-nv**
## Summary

- **Lightweight preview reads:** Move row-limiting (`nrows`) from `_run_internal` into `read_input` so that `preview()` reads only the requested number of rows from disk instead of loading the full dataset and truncating afterward. For CSV this uses `pd.read_csv(nrows=…)` natively; for Parquet it uses `pyarrow.parquet.ParquetFile.iter_batches()` to read only the necessary row batches, avoiding full-file deserialization. Handles `nrows ≤ 0` gracefully by returning a schema-preserving empty DataFrame for both formats.
- **Preserve downstream preview signal**: Keep `preview_num_records` threaded through `_run_internal` into detection, replacement, and rewrite workflows so that `NddAdapter` still routes to `DataDesigner.preview()` (in-memory execution) rather than `DataDesigner.create()` (artifact-backed execution). `run()` passes `None` to use full execution.
- **Clamp preview to available rows**: When `preview_num_records` exceeds the actual number of loaded rows, cap it to the available count and log a "capped" message. When it fits, log the processing count. This prevents downstream workflows from receiving a row count larger than the dataframe.
- **Simplify `_run_internal`**: Remove the old `effective_records` / `min(preview_num_records, num_records)` logic that lived further down the method — the dataframe is now already the right size on entry, and clamping happens up front if necessarily.
- **Move loading logs into `reader.py`**: The "Loaded N records" log now lives in `_load_dataframe`, with a "Preview mode" prefix when `nrows` is set.
- **Add unit tests**:
  - `read_input` `nrows` behavior: CSV truncation, Parquet truncation, `nrows > len`, `nrows=None`, attribute preservation, and remote-CSV forwarding.
  - Workflow-boundary tests: `preview_num_records` is capped and forwarded in `preview()`, forwarded as-is when smaller than available rows, and `None` in `run()`.
  - Logging assertions: verify preview-mode log appears, large-input preview only loads requested rows, and existing logging tests updated to match the new log format.

## Type of Change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update
- [ ] Refactoring

## Testing
- [x] `make test` passes locally
- [ ] `make check` passes locally (format + lint + typecheck + lock-check)
- [ ] Added/updated tests for changes

## Documentation
- [ ] If docs changed: `make docs-build` passes locally

### CODE REVIEW

File: src/anonymizer/interface/anonymizer.py [hunk 1]
```diff
@@ -143,7 +144,8 @@ def preview(
             num_records: Maximum records to process (default 10).
         """
         self._validate_preflight_config(config)
-        result = self._run_internal(config=config, data=data, preview_num_records=num_records)
+        input_df = read_input(data, nrows=num_records)
```
Comment (lipikaramaswamy) [inline src/anonymizer/interface/anonymizer.py:147]: Previously, `preview()` propagated `preview_num_records` through detection / replace / rewrite, which means `NddAdapter.run_workflow()` used `DataDesigner.preview(...)` rather than `DataDesigner.create(...)`.

After this refactor, preview reads fewer rows up front, but the downstream workflows no longer know they are in preview mode. In Data Designer 0.5.5, preview vs create is a real execution difference, not just a row-count difference:
  - `create()` goes through `builder.build(...)` / artifact-backed results
  - `preview()` goes through `builder.build_preview(...)` + `process_preview(...)` and stays in-memory

Our goal is to still keep the dataframe in memory for preview. So I'd recommend keeping the lightweight input read, but still preserve the downstream preview signal. So `preview()` can call `read_input(data, nrows=num_records)` and still pass `preview_num_records=num_records` through `_run_internal()`.

File: src/anonymizer/engine/io/reader.py [hunk 2]
```diff
@@ -32,15 +41,28 @@ def _validate_internal_column_collision(dataframe: pd.DataFrame, *, selected_tex
...
     if suffix not in SUPPORTED_IO_FORMATS:
         supported_formats = " or ".join(SUPPORTED_IO_FORMATS)
         raise InvalidInputError(f"Unsupported input format: {suffix}. Use {supported_formats}.")
     try:
         if suffix == ".csv":
-            return pd.read_csv(source_str)
-        return pd.read_parquet(source_str)
+            df = pd.read_csv(source_str, nrows=nrows)
+        else:
+            df = pd.read_parquet(source_str)
```
Comment (lipikaramaswamy) [inline src/anonymizer/engine/io/reader.py]: So csv loading is lightweight, but we don't actually get the same win for parquets. We'd need a true partial parquet reader (for example via `pyarrow.parquet.ParquetFile(...).iter_batches(...)`). If it's too complex, it's ok to leave out of scope for this PR. But let's clarify in the PR description, and perhaps a TODO that we can follow up on later.

File: src/anonymizer/interface/anonymizer.py (hunk 1 above)
Comment (lipikaramaswamy) [reply src/anonymizer/interface/anonymizer.py:147]: Also, let's add a test with one assertion at the workflow boundary that `Anonymizer.preview()` still routes through preview mode downstream (for example by verifying `preview_num_records` is forwarded to the detection workflow / adapter path). That would catch the semantic regression above.

Comment (memadi-nv) [reply src/anonymizer/interface/anonymizer.py:147]: Makes sense, Thanks for the feedback!

Comment (memadi-nv) [reply src/anonymizer/interface/anonymizer.py:147]: Added some edge cases as well. I might have overcomplicated the conditions. Please let me know your thoughts @lipikaramaswamy

File: src/anonymizer/engine/io/reader.py [hunk 3]
```diff
@@ -32,15 +43,42 @@ def _validate_internal_column_collision(dataframe: pd.DataFrame, *, selected_tex
     )
 
 
-def _load_dataframe(input_data: AnonymizerInput) -> pd.DataFrame:
+def _read_parquet_partial(source: str, *, nrows: int | None = None) -> pd.DataFrame:
```
Comment (lipikaramaswamy) [inline src/anonymizer/engine/io/reader.py:46]: One small edge case to consider here: `_read_parquet_partial()` does not currently guard `nrows <= 0` /
  empty-batch cases. It may be worth adding a tiny regression test for `nrows=0` (and/or empty parquet input) so this path  returns an empty dataframe cleanly instead of relying on `iter_batches(batch_size=0)` / `Table.from_batches([])` behavior.

**Review by lipikaramaswamy — APPROVED** [review]
LGTM. I tested with a few 10000+ record parquets and data loading was indeed faster.


One non-blocking test-structure nit: the new `nrows` coverage in `tests/engine/test_io.py` is solid, but there’s a fair amount of repeated file/data setup inline now. I’d consider extracting a small helper or fixture for "write input dataframe to csv/parquet and return an `AnonymizerInput`" so the tests stay focused on the `nrows` behavior instead of the file-construction boilerplate. Also, in `tests/interface/test_anonymizer_logging.py`, it would be cleaner to use `tmp_path` rather than the local `tempfile.mktemp(...)` helper.
